ReactNativeIsland.cpp

```diff
@@ -0,0 +1,113 @@
+#include "pch.h"
+#include "ReactNativeIsland.h"
+#include "ReactNativeIsland.g.cpp"
+#include <winrt/Microsoft.UI.Interop.h>
+
+namespace winrt::ReactNative
+{
+    using namespace winrt::Microsoft::ReactNative;
+    using namespace winrt::Microsoft::ReactNative::Composition;
+}
+
+namespace winrt::UI
+{
+    using namespace winrt::Microsoft::UI::Composition;
+    using namespace winrt::Microsoft::UI::Content;
+    using namespace winrt::Microsoft::UI::Input;
+    using namespace winrt::Microsoft::UI::Windowing;
+}
+
+float ScaleFactor(HWND hwnd) noexcept {
+    return GetDpiForWindow(hwnd) / static_cast<float>(USER_DEFAULT_SCREEN_DPI);
+}
+
+namespace winrt::Microsoft::ReactNative::implementation
+{
+    winrt::ReactNative::ReactNativeIsland ReactNativeIsland::Create(
+        winrt::UI::Compositor const& compositor,
+        winrt::UI::AppWindow const& window,
+        winrt::ReactNative::ReactNativeHost const& host,
+        winrt::hstring const& mainComponentName)
+    {
+        return winrt::make<ReactNativeIsland>(compositor, window, host, mainComponentName);
+    }
+
+    ReactNativeIsland::~ReactNativeIsland()
+    {
+        // Does the CompositionRootView close the Island? Or do we close the island independently?
+        m_island.Close();
+    }
+
+    void ReactNativeIsland::SetScaleFactor(winrt::UI::AppWindow const& window)
+    {
+        auto hwnd = winrt::Microsoft::UI::GetWindowFromWindowId(window.Id());
+        m_scaleFactor = ScaleFactor(hwnd);
+    }
+
+    ReactNativeIsland::ReactNativeIsland(
+        winrt::UI::Compositor const& compositor,
+        winrt::UI::AppWindow const& window,
+        winrt::ReactNative::ReactNativeHost const& host,
+        winrt::hstring const& mainComponentName)
+        : m_compositor(compositor)
+    {
+        SetScaleFactor(window);
+
+        winrt::ReactNative::ReactViewOptions viewOptions;
+        viewOptions.ComponentName(mainComponentName);
+
+        // Create a RootView (creates a spriteVisual) and associates it with an island
+        m_rootView = winrt::ReactNative::CompositionRootView(m_compositor);
+        m_rootView.ReactViewHost(
+            winrt::ReactNative::ReactCoreInjection::MakeViewHost(host, viewOptions));
+
+        // Create a contentIsland
+        m_island = m_rootView.Island();
+
+        auto invScale = 1.0f / m_scaleFactor;
+        m_rootView.RootVisual().Scale({ invScale, invScale, invScale });
+        m_rootView.ScaleFactor(m_scaleFactor);
```

Comment:
Is `SetScaleFactor()` ever expected to be called a second time? If so, wouldn't this need to be updated too?